helix-core/src/main/java/org/apache/helix/manager/zk/zookeeper/ZkClient.java

```diff
@@ -1261,31 +1257,75 @@ private void processDataOrChildChange(WatchedEvent event, long notificationTime)
 
     if (event.getType() == EventType.NodeDataChanged || event.getType() == EventType.NodeDeleted
         || event.getType() == EventType.NodeCreated) {
+      boolean pathExists = event.getType() != EventType.NodeDeleted;
       Set<IZkDataListenerEntry> listeners = _dataListener.get(path);
       if (listeners != null && !listeners.isEmpty()) {
-        fireDataChangedEvents(event.getPath(), listeners, OptionalLong.of(notificationTime));
+        fireDataChangedEvents(event.getPath(), listeners, notificationTime, pathExists);
+      }
+    }
+  }
+
+  /**
+   * The method is called when state change occurs and we need to handle existing path data handlers
+   */
+  private void fireDataChangedEvents(final String path, Set<IZkDataListenerEntry> listeners) {
+    try {
+      for (final IZkDataListenerEntry listener : listeners) {
+        ZkEvent zkEvent = new ZkEvent(
+            "Data of " + path + " changed sent to " + listener.getDataListener()
+                + " prefetch data: " + listener.isPrefetchData()) {
+          @Override
+          public void run()
+              throws Exception {
+            Object data;
+            // TODO: are we fetching the data multiple times?
```

Comment:
The answer is yes. There is a solution, can we do the same as the other methods so as to avoid multiple reads?

Reply:
I will make it a formal TODO and let's first focus on the watcher leakage issue :)